### PR TITLE
Add AMP for TF Albert

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?

This PR adds the following features to TF Albert:

- AMP compliancy
- Loss computation for TFAlbertForPreTraining
- Cleaning source code

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Looks good to me, thanks for fixing!

File: src/transformers/models/bert/modeling_tf_bert.py [hunk 1]
```diff
@@ -1009,7 +1008,8 @@ def call(
             total_loss = self.compute_loss(labels=d_labels, logits=(prediction_scores, seq_relationship_score))
 
         if not inputs["return_dict"]:
-            return (prediction_scores, seq_relationship_score) + outputs[2:]
+            output = (prediction_scores, seq_relationship_score) + outputs[2:]
+            return ((total_loss,) + output) if total_loss is not None else output
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_tf_bert.py:1012]: It was not returning the loss before that?

Comment (LysandreJik) [reply src/transformers/models/bert/modeling_tf_bert.py:1012]: This should have been caught by a test

File: tests/test_modeling_tf_albert.py [hunk 2]
```diff
@@ -242,6 +243,16 @@ class TFAlbertModelTest(TFModelTesterMixin, unittest.TestCase):
...
+    # special case for ForPreTraining model
+    def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
+        inputs_dict = super()._prepare_for_class(inputs_dict, model_class, return_labels=return_labels)
+
+        if return_labels:
+            if model_class in TF_MODEL_FOR_PRETRAINING_MAPPING.values():
+                inputs_dict["sentence_order_label"] = tf.zeros(self.model_tester.batch_size, dtype=tf.int32)
+
+        return inputs_dict
+
```
Comment (LysandreJik) [inline tests/test_modeling_tf_albert.py:255]: I see it will get caught by a test from now on :)

**Review by LysandreJik — COMMENTED** [review]
Hi! Could you split the two objectives of cleaning the source code and AMP in two PRs? It's better because:

- It will be easier to detect the issue if we identify an issue with this commit down the road
- It makes it easier to read. I can't see what changes are done here related to AMP

File: src/transformers/models/bert/modeling_tf_bert.py (hunk 1 above)
Comment (jplu) [reply src/transformers/models/bert/modeling_tf_bert.py:1012]: The test on the loss computation was working because the test is run only for `return_dict=True`, it would have failed if `return_dict=False`.

File: src/transformers/models/albert/modeling_tf_albert.py [hunk 3]
```diff
@@ -150,107 +187,95 @@ def call(
...
         query_layer = self.transpose_for_scores(mixed_query_layer, batch_size)
         key_layer = self.transpose_for_scores(mixed_key_layer, batch_size)
         value_layer = self.transpose_for_scores(mixed_value_layer, batch_size)
 
         # Take the dot product between "query" and "key" to get the raw attention scores.
         # (batch size, num_heads, seq_len_q, seq_len_k)
         attention_scores = tf.matmul(query_layer, key_layer, transpose_b=True)
-        # scale attention_scores
-        dk = tf.cast(shape_list(key_layer)[-1], tf.float32)
-        attention_scores = attention_scores / tf.math.sqrt(dk)
```
Comment (jplu) [inline src/transformers/models/albert/modeling_tf_albert.py:223]: First fix for AMP, (removing the cast into `tf.float32`)

File: src/transformers/models/albert/modeling_tf_albert.py [hunk 4]
```diff
@@ -550,9 +601,10 @@ def call(
         # positions we want to attend and -10000.0 for masked positions.
         # Since we are adding it to the raw scores before the softmax, this is
         # effectively the same as removing these entirely.
-
-        extended_attention_mask = tf.cast(extended_attention_mask, tf.float32)
```
Comment (jplu) [inline src/transformers/models/albert/modeling_tf_albert.py:554]: Second fix for AMP, still remove the cast into `tf.float32`

**Comment by jplu** [comment]
I can split this PR into two different ones, but the one on AMP will be very short (only two single line to update, see the review above). Are you agree with a that tiny PR? If it is still ok, I will split this one^^

**Comment by LysandreJik** [comment]
It's ok, thanks for showing me the changes!

**Review by LysandreJik — APPROVED** [review]
The changes LGTM

**Comment by jplu** [comment]
@patrickvonplaten feel free to merge if it looks ok for you!

**Review by patrickvonplaten — APPROVED** [review]
